roachtest: json_object_agg crash

[(roachtest).sqlsmith/setup=seed/setting=no-mutations failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2283744&tab=buildLog) on [release-20.2@4ecea82f7dfaf5a4edd4d7c7ffad4a4e7c6c4cce](https://github.com/cockroachdb/cockroach/commits/4ecea82f7dfaf5a4edd4d7c7ffad4a4e7c6c4cce):

```
The test failed on branch=release-20.2, cloud=gce:
test artifacts and logs in: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/sqlsmith/setup=seed/setting=no-mutations/run_1
	sqlsmith.go:177,sqlsmith.go:202,test_runner.go:754: error: pq: communication error: rpc error: code = Canceled desc = context canceled
		stmt:
		SELECT
			json_object_agg(tab_14985._string::STRING, tab_14985._string) OVER (PARTITION BY tab_14985.tableoid ORDER BY tab_14985._jsonb DESC, tab_14985._string DESC, tab_14985._timestamptz DESC)::JSONB
				AS col_37213,
			max(tab_14985._timestamptz::TIMESTAMPTZ)::TIMESTAMPTZ AS col_37214,
			variance(tab_14985.crdb_internal_mvcc_timestamp::DECIMAL)::DECIMAL AS col_37215
		FROM
			defaultdb.public.seed@seed__int8__float8__date_idx AS tab_14985
		WHERE
			tab_14985._bool
		GROUP BY
			tab_14985._bytes,
			tab_14985.tableoid,
			tab_14985._string,
			tab_14985._jsonb,
			tab_14985.crdb_internal_mvcc_timestamp,
			tab_14985._interval,
			tab_14985._timestamptz
		LIMIT
			40:::INT8;

	cluster.go:1651,context.go:135,cluster.go:1640,test_runner.go:823: dead node detection: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/bin/roachprod monitor teamcity-2283744-1600497107-13-n4cpu4 --oneshot --ignore-empty-nodes: exit status 1 3: dead
		4: 7346
		2: 7565
		1: 9499
		Error: UNCLASSIFIED_PROBLEM: 3: dead
		(1) UNCLASSIFIED_PROBLEM
		Wraps: (2) attached stack trace
		  -- stack trace:
		  | main.glob..func14
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachprod/main.go:1143
		  | main.wrap.func1
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachprod/main.go:267
		  | github.com/spf13/cobra.(*Command).execute
		  | 	/home/agent/work/.go/pkg/mod/github.com/spf13/cobra@v0.0.5/command.go:830
		  | github.com/spf13/cobra.(*Command).ExecuteC
		  | 	/home/agent/work/.go/pkg/mod/github.com/spf13/cobra@v0.0.5/command.go:914
		  | github.com/spf13/cobra.(*Command).Execute
		  | 	/home/agent/work/.go/pkg/mod/github.com/spf13/cobra@v0.0.5/command.go:864
		  | main.main
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachprod/main.go:1839
		  | runtime.main
		  | 	/usr/local/go/src/runtime/proc.go:203
		  | runtime.goexit
		  | 	/usr/local/go/src/runtime/asm_amd64.s:1357
		Wraps: (3) 3: dead
		Error types: (1) errors.Unclassified (2) *withstack.withStack (3) *errutil.leafError

```

<details><summary>More</summary><p>


Artifacts: [/sqlsmith/setup=seed/setting=no-mutations](https://teamcity.cockroachdb.com/viewLog.html?buildId=2283744&tab=artifacts#/sqlsmith/setup=seed/setting=no-mutations)

[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2Asqlsmith%2Fsetup%3Dseed%2Fsetting%3Dno-mutations.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

```
panic: modify after Build()

goroutine 40615 [running]:
panic(0x4cae060, 0xc000b19f90)
	/usr/local/go/src/runtime/panic.go:722 +0x2c2 fp=0xc0023b18c8 sp=0xc0023b1838 pc=0x7b1aa2
github.com/cockroachdb/cockroach/pkg/util/json.(*ObjectBuilder).Add(0xc0019405c0, 0x8b9c4b3, 0x1, 0x631ab80, 0xc000b19f70)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/json/json.go:299 +0x15f fp=0xc0023b1918 sp=0xc0023b18c8 pc=0x17bfd3f
github.com/cockroachdb/cockroach/pkg/util/json.(*ObjectBuilderWithCounter).Add(0xc000b19f30, 0x8b9c4b3, 0x1, 0x631ab80, 0xc000b19f70)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/json/json.go:267 +0x5a fp=0xc0023b1950 sp=0xc0023b1918 pc=0x17bfb1a
github.com/cockroachdb/cockroach/pkg/sql/sem/builtins.(*jsonObjectAggregate).Add(0xc00256ff50, 0x61a2700, 0xc0026f32c0, 0x6302ca0, 0xc002617a10, 0xc001940730, 0x1, 0x1, 0x0, 0xc002617c00)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/sem/builtins/aggregate_builtins.go:3605 +0xf4 fp=0xc0023b19b0 sp=0xc0023b1950 pc=0x1d3a3f4
github.com/cockroachdb/cockroach/pkg/sql/sem/builtins.(*aggregateWindowFunc).Compute(0xc00256ffb0, 0x61a2700, 0xc0026f32c0, 0xc0028f18c0, 0xc0019362d0, 0x0, 0x0, 0x3, 0xc0023b1ab8)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/sem/builtins/window_builtins.go:298 +0x1d2 fp=0xc0023b1a28 sp=0xc0023b19b0 pc=0x1d5b5c2
github.com/cockroachdb/cockroach/pkg/sql/sem/builtins.(*framableAggregateWindowFunc).Compute(0xc001962480, 0x61a2700, 0xc0026f32c0, 0xc0028f18c0, 0xc0019362d0, 0xc0019406a0, 0x0, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/sem/builtins/window_builtins.go:384 +0x5b2 fp=0xc0023b1ac8 sp=0xc0023b1a28 pc=0x1d5be32
github.com/cockroachdb/cockroach/pkg/sql/rowexec.(*windower).processPartition(0xc000d2bb00, 0x61a2700, 0xc0026f32c0, 0xc0028f18c0, 0xc002555500, 0x0, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/rowexec/windower.go:583 +0x2ee fp=0xc0023b1ce8 sp=0xc0023b1ac8 pc=0x23ae6be
github.com/cockroachdb/cockroach/pkg/sql/rowexec.(*windower).computeWindowFunctions(0xc000d2bb00, 0x61a2700, 0xc0026f32c0, 0xc0028f18c0, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/rowexec/windower.go:712 +0xb85 fp=0xc0023b1e48 sp=0xc0023b1ce8 pc=0x23b0a55
github.com/cockroachdb/cockroach/pkg/sql/rowexec.(*windower).emitRow(0xc000d2bb00, 0x2, 0x0, 0x0, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/rowexec/windower.go:326 +0x9b fp=0xc0023b1e98 sp=0xc0023b1e48 pc=0x23adb0b
github.com/cockroachdb/cockroach/pkg/sql/rowexec.(*windower).Next(0xc000d2bb00, 0x0, 0x0, 0x0, 0xc0029e1b38)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/rowexec/windower.go:236 +0x54 fp=0xc0023b1f00 sp=0xc0023b1e98 pc=0x23ad464
github.com/cockroachdb/cockroach/pkg/sql/execinfra.Run(0x61a2700, 0xc0026f32c0, 0x61d0780, 0xc000d2bb00, 0x617a940, 0xc001a3f880)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/execinfra/base.go:170 +0x35 fp=0xc0023b1f40 sp=0xc0023b1f00 pc=0x2134ce5
github.com/cockroachdb/cockroach/pkg/sql/execinfra.(*ProcessorBase).Run(0xc000d2bb00, 0x61a2700, 0xc0026f32c0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/execinfra/processorsbase.go:775 +0x92 fp=0xc0023b1f80 sp=0xc0023b1f40 pc=0x213b9f2
github.com/cockroachdb/cockroach/pkg/sql/flowinfra.(*FlowBase).startInternal.func1(0xc0026f33c0, 0x1, 0x4, 0x61a2700, 0xc0026f32c0, 0xc001d8a120, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/flowinfra/flow.go:326 +0x5c fp=0xc0023b1fa8 sp=0xc0023b1f80 pc=0x23585fc
runtime.goexit()
	/usr/local/go/src/runtime/asm_amd64.s:1357 +0x1 fp=0xc0023b1fb0 sp=0xc0023b1fa8 pc=0x7e2271
created by github.com/cockroachdb/cockroach/pkg/sql/flowinfra.(*FlowBase).startInternal
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/flowinfra/flow.go:325 +0x2d9
```

--- a/pkg/sql/logictest/testdata/logic_test/window
+++ b/pkg/sql/logictest/testdata/logic_test/window
@@ -3838,3 +3838,27 @@ statement ok
 SELECT max(b::INT8) OVER (PARTITION BY b ORDER BY a RANGE 1 PRECEDING)
 FROM t53442_b NATURAL JOIN t53442_a
 WHERE false
+
+# Regression test for a crash with json_object_agg used as a window function
+# (#54604).
+statement ok
+CREATE TABLE t54604 (s STRING);
+INSERT INTO t54604 SELECT g::STRING FROM generate_series(1, 5) AS g
+
+query T rowsort
+SELECT json_object_agg(s, s) OVER (ORDER BY s DESC) FROM t54604
+----
+{"1": "1", "2": "2", "3": "3", "4": "4", "5": "5"}
+{"2": "2", "3": "3", "4": "4", "5": "5"}
+{"3": "3", "4": "4", "5": "5"}
+{"4": "4", "5": "5"}
+{"5": "5"}
+
+query T rowsort
+SELECT jsonb_object_agg(s, s) OVER (ORDER BY s RANGE UNBOUNDED PRECEDING) FROM t54604
+----
+{"1": "1"}
+{"1": "1", "2": "2"}
+{"1": "1", "2": "2", "3": "3"}
+{"1": "1", "2": "2", "3": "3", "4": "4"}
+{"1": "1", "2": "2", "3": "3", "4": "4", "5": "5"}

--- a/pkg/sql/rowexec/windower.go
+++ b/pkg/sql/rowexec/windower.go
@@ -449,6 +449,10 @@ func (w *windower) processPartition(
 	for _, windowFnIdx := range w.orderOfWindowFnsProcessing {
 		windowFn := w.windowFns[windowFnIdx]
 
+		// TODO(yuzefovich): creating a new WindowFrameRun object for each
+		// partition and populating it below for a custom window frame is
+		// suboptimal. Consider extracting this logic into the constructor of
+		// the windower and reusing the same objects between partitions.
 		frameRun := &tree.WindowFrameRun{
 			ArgsIdxs:     windowFn.argsIdxs,
 			FilterColIdx: windowFn.filterColIdx,

--- a/pkg/sql/sem/builtins/window_builtins.go
+++ b/pkg/sql/sem/builtins/window_builtins.go
@@ -342,9 +342,16 @@ type framableAggregateWindowFunc struct {
 func newFramableAggregateWindow(
 	agg tree.AggregateFunc, aggConstructor func(*tree.EvalContext, tree.Datums) tree.AggregateFunc,
 ) tree.WindowFunc {
+	// jsonObjectAggregate is a special aggregate function because its
+	// implementation assumes that once Result is called, the returned
+	// object is immutable and calls to Add will result in a panic. To go
+	// around this limitation, we make sure that the function is reset for
+	// each row regardless of the window frame.
+	_, shouldReset := agg.(*jsonObjectAggregate)
 	return &framableAggregateWindowFunc{
 		agg:            &aggregateWindowFunc{agg: agg, peerRes: tree.DNull},
 		aggConstructor: aggConstructor,
+		shouldReset:    shouldReset,
 	}
 }
 

